src/hotspot/share/runtime/thread.inline.hpp

```diff
@@ -135,24 +135,34 @@ inline JavaThreadState JavaThread::thread_state() const    {
 #if defined(PPC64) || defined (AARCH64)
```

Comment:
Should it still be platform dependent? The headline suggest no.

Reply:
As I tried to make clear, I made no changes to the default/existing methods. Instead I added specific acquire/release variants and used them where it was (fairly) obvious they needed to be used.

I think I need to adjust the issue title as initially the thought was that acquire/release would be needed in many places, but it turned out not to be the case.

Reply:
The platform dependency looks pointless after your change. Aren't Atomic::release_store and Atomic::store equivalent on TSO platforms (same for Atomic::load_acquire and Atomic::load)?

Reply:
I don't see how it looks pointless. I agree it is not clean - but that arose because the PPC64 port decided that accessing thread-state always required acquire/release semantics even if not actually the case. There are cases when acquire/release (or other semantics are needed) and in those existing cases we ended up with redundancy by having an external storestore as well as the release semantics. That has now been fixed. If the PPC64 folk were to look at the remaining "plain" accesses of thread-state they may decide the platform dependency is not needed, but that is up to them.

I don't get your point about the equivalence of operations on TSO. It doesn't matter if operations are equivalent under the hood, we should always use the operation that expresses the desired semantics - at least in shared code.

Reply:
The different implementation for PPC64 was introduced by
https://bugs.openjdk.java.net/browse/JDK-8029396
which is part of the initial PPC64 port. We didn't have clean platform independent code in Hotspot at that point of time.
It was figured out that other non-TSO platforms need memory barriers as well:
https://bugs.openjdk.java.net/browse/JDK-8069593
So it looks like using release/acquire semantics is what's actually needed in general. Seems like nobody cared about TSO platform's implementation because there are obviously no real problems.

Reply:
I don't agree with the assertion we didn't have clean platform independent code when PPC64 was introduced as justification here. We already had closed ports of ARM and PPC at that time and we had done a lot (ever since IA64 work years earlier) to make the shared code platform (and TSO) agnostic. Sure it wasn't/isn't perfect but I don't think you can use that as a justification here. I couldn't find anything that explained why PPC64 took the path it did (nor the discussion that forced the acquire/release to be restricted to PPC64 due to performance concerns on other platforms). The Aarch64 port at least tries to justify it, but I'd argue always using acquire/release semantics with thread-state was a sledgehammer approach to avoid actually identifying which interactions actually need what kind of memory barriers.

Reply:
I'm not happy with the sledgehammer, either, but better using a sledgehammer than having possibly incorrect code or an unstable VM IMHO.
No problems reported on some platforms doesn't mean that the code is correct. Do we know that not using release/acquire is correct for any platform? The aarch64 issue doesn't sound like that.
IA64 is not affected because volatile accesses always use release/acquire semantics. I believe that arm32 uses that, too. So, we already rely on release/acquire semantics. Shouldn't that better be explicit?

Reply:
The need for acquire/release semantics needs to be established for pairs of accesses. Simply using them "just in case" is not something I support and I certainly discourage it because it says "we don't understand our code so we're just throwing barriers in just in case".

That said there seems no interest in this change so I will just close it out.

Cheers,
David